### PR TITLE
[AutoModel] Split AutoModelWithLMHead into clm, mlm, encoder-decoder

### PR DESCRIPTION
**Pull request by patrickvonplaten**
This is a follow-up PR of #4874.

In #4874, `BertModelForMaskedLM` , was split into a clm `BertLMHeadModel` and mlm Bert Model `BertForMaskedLM`. In order for the encoder-decoder framework to work correctly, the `BertLMHeadModel` needs to be loaded, when instantiating an `EncoderDecoder.from_encoder_decoder_pretrained()`, therefore a new `AutoModelForCausalLM` has to be created.

This PR deprecates `AutoModelWithLMHead` and introduces:
- `AutoModelForCausalLM` for Autoregressive models
- `AutoModelForMaskedLM` for Autoencoding models
- `AutoModelForSeq2SeqCausalLM` for Sequence-to-sequence models with causal LM for the decoder

@julien-c @LysandreJik @sgugger @thomwolf @sshleifer -> `AutoModelWithLMHead` still works as before so no breaking changes, but it's deprecated. All AutoModels are exposed in the `__init__` (don't really see a reason why they shouldn't). What do you guys think about the naming?

**IMPORTANT:**
#4874 and this PR might introduce some breaking changes for the encoder-decoder framework:
Instead of using `AutoModelWithLMHead` one has to use `AutoModelForCausalLM` for the decoder model now and instead of using `BertForMaskedLM` one should use `BertLMHeadModel` from now one. 
There are **no breaking changes** for the encoder-decoder user-facing functions `.from_pretrained()` and `.from_encoder_decoder_pretrained()`

### CODE REVIEW

**Comment by sgugger** [comment]
This all looks good to me. I'd just simplify `AutoModelForSeq2SeqCausalLM` to `AutoModelForSeq2SeqLM` and the related names since I don't think there exists an encoder/decoder setup for LM where the decoder doesn't have a causal mask, so causal doesn't really add any needed info to the name.

File: src/transformers/modeling_auto.py
Comment (sgugger) [inline src/transformers/modeling_auto.py:225]: Eh, I hadn't realized but `XLMWithLMHead` does not compute the loss when provided with labels because of this. Definitely needs a split and both splits to return the proper loss if provided with labels

File: src/transformers/modeling_auto.py
Comment (sgugger) [inline src/transformers/modeling_auto.py:774]: AutoModelWithLMHead -> AutoModelForCausalLM here and at line 776.

File: src/transformers/modeling_auto.py
Comment (sgugger) [inline src/transformers/modeling_auto.py:1050]: AutoModelWithLMHead -> AutoModelForSeq2SeqCausalLM

**Review by sgugger — APPROVED** [review]
Some minor edits needed but otherwise LGTM.

File: src/transformers/modeling_auto.py
Comment (LysandreJik) [inline src/transformers/modeling_auto.py:667]: I think it should be a `FutureWarning` rather than a `DeprecationWarning` so that users are aware of this change (`DeprecationWarnings` are ignored by default)

**Review by LysandreJik — APPROVED** [review]
Looks great!

**Comment by julien-c** [comment]
I like this but if we merge this, I think we should make a note to actually remove the `AutoModelWithLMHead` at the next (or next-next?) major release otherwise users will still use it and it will be confusing

**Comment by patrickvonplaten** [comment]
Merging to unblock the encoder decoder framework to work correctly. Pinging @thomwolf for notification.

**Comment by thomwolf** [comment]
Yes, I'm happy with this!

**Comment by lcyuanjiang** [comment]
what is the different when i pass a ones.tril() mask to the bertencoder

**Pull request merged**
